File: ios/EthereumCoin.swift
//
//  EthereumCoin.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class EthereumCoin: CoinFactory.Coin {
    private var address: String = ""
    
    override init(mnemonic: String) {
        super.init(mnemonic: mnemonic)
        self.address = CoinType.ethereum.deriveAddress(privateKey: wallet.getKeyForCoin(coin: .ethereum))
    }
    
    override func getNewAddress() -> String {
        return self.address
    }
    
    override func getPrivateKey() -> String {
        let privateKeyBytes = wallet.getKeyForCoin(coin: .ethereum).data
        return String(data: privateKeyBytes, encoding: .utf8) ?? ""
    }
    
    override func signTransaction(rawData: String) -> String {
        // Implement the logic to sign an Ethereum transaction using the provided raw data
        return ""
    }
}

File: android/app/src/main/java/com/coinswallet/coins/EthereumCoin.java
// EthereumCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;

import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.PrivateKey;
import java.util.Base64;
import java.nio.charset.StandardCharsets;
import android.util.Log;

public class EthereumCoin extends CoinFactory.Coin {
    private final String address;

    // static {
    // Log.d("coinswallet", "Registering EthereumCoin");
    // CoinFactory.registerCoin("ethereum", EthereumCoin::new);
    // }

    public EthereumCoin(String mnemonic) {
        super(mnemonic);
        this.address = CoinType.ETHEREUM.deriveAddress(wallet.getKeyForCoin(CoinType.ETHEREUM));
    }

    @Override
    public String getNewAddress() {
        return this.address;
    }

    @Override
    public String getPrivateKey() {
        byte[] privateKeyBytes = wallet.getKeyForCoin(CoinType.ETHEREUM).data();
        return new String(privateKeyBytes, StandardCharsets.UTF_8);
    }

    @Override
    public String signTransaction(String rawData) {
        return null;
    }
}
